packages/core/src/services/chains/run.ts

```diff
@@ -210,20 +226,42 @@ async function runStep({
       output: step.output,
     }).then((r) => r.unwrap())
 
+    const checkResult = checkValidType(aiResult)
+    if (checkResult.error) throw checkResult.error
+
     const consumedStream = await consumeStream({
       controller,
       result: aiResult,
     })
-    const response = await providerProcessor
-      .call({
-        aiResult,
-        finishReason: consumedStream.finishReason,
-        startTime: stepStartTime,
-      })
-      .then((r) => r.unwrap())
-
     if (consumedStream.error) throw consumedStream.error
 
+    const response = await processResponse({
+      aiResult,
+      apiProvider: step.provider,
+      config: step.config,
+      errorableUuid,
+      messages: step.conversation.messages,
+      source,
+      workspace,
+      startTime: stepStartTime,
+    })
+
+    const providerLog = await saveOrPublishProviderLogs({
+      workspace,
+      streamType: aiResult.type,
+      finishReason: consumedStream.finishReason,
+      data: buildProviderLogDto({
+        workspace,
+        source,
+        provider: step.provider,
+        conversation: step.conversation,
+        stepStartTime,
+        errorableUuid,
+        response,
+      }),
+      saveSyncProviderLogs: step.chainCompleted,
+    })
```

Comment:
The thing I don't like about not using classes is that force to use more functions. And make this entry point more complex. `run.ts` should be super easy to read and understand. Is a key piece. 

That was my reasoning when I encapsulated some pieces in classes.

Reply:
hmm the file should be exactly the same length or less since changing the implementation from classes to functional does not change the amount of calls. In fact it should be simpler since this:

```
validateChain(...)
```

is simpler to parse than:
```
const chainValidator = new ChainValidator(...)

chainValidator.call()
```

If you are seeing more function calls is because I've decoupled some code that was previously coupled (because I need it to not be coupled to fix the bug), mainly ProviderProcessor which was doing more than I needed it to do